src/client/components/scroller/scroller.tsx

```diff
@@ -2,44 +2,380 @@ require('./scroller.css');
+  layout: ScrollerLayout;
+
+  onClick?: (x: number, y: number) => void;
+  onMouseMove?: (x: number, y: number) => void;
+  onMouseLeave?: () => void;
+  onScroll?: (scrollTop: number, scrollLeft: number) => void;
+
+  // Transcluded elements
+  topGutter?: JSX.Element | JSX.Element[];
+  rightGutter?: JSX.Element | JSX.Element[];
+  bottomGutter?: JSX.Element | JSX.Element[];
+  leftGutter?: JSX.Element | JSX.Element[];
+  topLeftCorner?: JSX.Element | JSX.Element[];
+  topRightCorner?: JSX.Element | JSX.Element[];
+  bottomRightCorner?: JSX.Element | JSX.Element[];
+  bottomLeftCorner?: JSX.Element | JSX.Element[];
+  body?: JSX.Element[];
+  bodyOverlay?: JSX.Element | JSX.Element[];
 }
 
 export interface ScrollerState {
-  scrollLeft: number;
-  scrollTop: number;
+  scrollTop?: number;
+  scrollLeft?: number;
+
+  viewportHeight?: number;
+  viewportWidth?: number;
 }
 
 export class Scroller extends React.Component<ScrollerProps, ScrollerState> {
-
   constructor() {
     super();
     this.state = {
+      scrollTop: 0,
       scrollLeft: 0,
-      scrollTop: 0
+      viewportHeight: 0,
+      viewportWidth: 0
+    };
+  }
+
+  getTopLeftCornerStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    return {
+      width: layout.left,
+      height: layout.top,
+      left: 0
+    };
+  }
+
+  getTopRightCornerStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    return {
+      width: layout.right,
+      height: layout.top,
+      right: 0
+    };
+  }
+
+  getBottomLeftCornerStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    return {
+      width: layout.left,
+      height: layout.bottom,
+      bottom: 0
+    };
+  }
+
+  getBottomRightCornerStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    return {
+      width: layout.right,
+      height: layout.bottom,
+      bottom: 0,
+      right: 0
+    };
+  }
+
+  getTopGutterStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    const { scrollLeft } = this.state;
+
+    return {
+      height: layout.top,
+      left: layout.left - scrollLeft,
+      right: layout.right
+    };
+  }
+
+  getRightGutterStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    const { scrollTop } = this.state;
+
+    return {
+      width: layout.right,
+      right: 0,
+      top: layout.top - scrollTop,
+      bottom: layout.bottom
+    };
+  }
+
+  getBottomGutterStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    const { scrollLeft } = this.state;
+
+    return {
+      height: layout.bottom,
+      left: layout.left - scrollLeft,
+      right: layout.right
+    };
+  }
+
+  getLeftGutterStyle(): React.CSSProperties {
+    const { layout } = this.props;
+    const { scrollTop } = this.state;
+
+    return {
+      width: layout.left,
+      left: 0,
+      top: layout.top - scrollTop,
+      bottom: layout.bottom
+    };
+  }
+
+  getTopShadowStyle(): React.CSSProperties {
```

Comment:
similarly I would write this as `getShadowStyle(side: Side): React.CSSProperties` and have a 4 way switch statement inside.